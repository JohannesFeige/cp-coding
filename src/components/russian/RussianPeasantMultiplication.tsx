import React, { useCallback, useState } from 'react';
import { mul } from '../shared/calculation.service';

interface State {
  x: number;
  y: number;
  result: number;
}

const RussianPeasantMultiplication: React.FC = () => {
  const [state, setState] = useState<State>({ x: 0, y: 0, result: 0 });
  const submitFormHandler = useCallback((event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
  }, []);

  const xChangeHandler = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const newX = +event.target.value;
      setState({ ...state, x: newX, result: mul(newX, state.y) });
    },
    [state]
  );

  const yChangeHandler = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const newY = +event.target.value;
      setState({ ...state, y: newY, result: mul(state.x, newY) });
    },
    [state]
  );

  return (
    <React.Fragment>
      <form onSubmit={submitFormHandler}>
        <div className="form-row">
          <div className="form-group col-md-2">
            <input type="number" className="form-control" placeholder="X" value={state.x} onChange={xChangeHandler} />
          </div>
          *
          <div className="form-group col-md-2">
            <input type="number" className="form-control" placeholder="Y" value={state.y} onChange={yChangeHandler} />
          </div>
          =
          <div className="form-group col-md-2">
            <span className="form-control" id="result">
              {state.result}
            </span>
          </div>
        </div>
      </form>
    </React.Fragment>
  );
};

export default RussianPeasantMultiplication;
