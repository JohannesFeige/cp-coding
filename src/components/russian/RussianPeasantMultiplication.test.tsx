import React from 'react';
import { shallow, mount } from 'enzyme';
import RussianPeasantMultiplication from './RussianPeasantMultiplication';

describe('RussianPeasantMultiplication', () => {
  it('should render', () => {
    const wrapper = shallow(<RussianPeasantMultiplication />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render and calculate', () => {
    const wrapper = mount(<RussianPeasantMultiplication />);
    const xInput = wrapper.find('input[type="number"][placeholder="X"]');
    expect(xInput.getElements().length).toBe(1);
    const yInput = wrapper.find('input[type="number"][placeholder="Y"]');
    expect(yInput.getElements().length).toBe(1);
    xInput.simulate('change', {
      target: {
        value: 42,
      },
    });
    yInput.simulate('change', {
      target: {
        value: 47,
      },
    });
    const result = wrapper.find('span');
    expect(result.getElements()).toHaveLength(1);
    expect(result.first().text()).toBe('1974');
  });
});
