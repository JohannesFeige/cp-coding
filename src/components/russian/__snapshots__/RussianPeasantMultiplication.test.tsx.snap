// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RussianPeasantMultiplication should render 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="form-group col-md-2"
      >
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="X"
          type="number"
          value={0}
        />
      </div>
      *
      <div
        className="form-group col-md-2"
      >
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Y"
          type="number"
          value={0}
        />
      </div>
      =
      <div
        className="form-group col-md-2"
      >
        <span
          className="form-control"
          id="result"
        >
          0
        </span>
      </div>
    </div>
  </form>
</Fragment>
`;
